GH-1333: Implement Screen 3 and 4 of v6.0 navigation and FTUE

hahmadia opened the pull request:
Implement Screen 3 and 4 of v6.0 navigation and FTUE 

Fixes #1333 

Implementation details can be https://docs.google.com/document/d/1Q0eujL89EU39Z3R1mer-tZMksEbyLtPGNqO-JsAe9Gg/edit?usp=sharing

hahmadia commented:
Hi @asaadmahmood ,

I've done what I think to be the bulk of the CSS work required for this PR. I've just added you to double check my work and make any adjustments you feel are needed to be made. 

Thanks

sbishel commented on webapp/src/components/emptyCenterPanel.tsx:
```diff
@@ -1,34 +1,170 @@
+            {!readonly &&
+                <BoardTemplateMenuItemOptionsButton
+                    showBoard={showBoard!}
+                    boardTemplate={boardTemplate!}
```
I find this to be very confusing. The "!" tells the compiler that this variable is NOT null or undefined, yet there is no other check to verify that and there are definitely instances these are undefined.  I'm actually surprised this works, I had to try it.  I guess the React simply doesn't render if a required property is null.

In any case, it may be better explicitly checking for those variables first. 

```
{!readonly && showBoard && boardTemplate &&
     <BoardTemplateMenuItemOptionsButton
     ...
```

sbishel commented on webapp/src/components/sidebar/boardTemplateMenuItem.tsx:
```diff
@@ -39,8 +32,49 @@ const addBoardFromTemplate = async (intl: IntlShape, showBoard: (id: string) =>
+    boardTemplate: Board
+}
+
+export const BoardTemplateMenuItemOptionsButton = React.memo((props: ButtonProps) => {
```
Can this name be simplified. `BoardTemplateButtonMenu`

sbishel commented on webapp/src/components/workspaceSwitcher/workspaceSwitcher.tsx:
```diff
@@ -56,6 +57,14 @@ const WorkspaceSwitcher = (props: Props): JSX.Element => {
+            {activeWorkspace &&
+                <span
+                    className='add-workspace-icon'
+                    onClick={() => history.push(`/workspace/${activeWorkspace?.id}?showEmptyCenterPanel=true`)}
```
I don't like using a query string for this. Can we just send it to the workspace? I realize we have code in `boardPage.tsx` that loads the lastBoard and lastView if not passed. I'm in favor of either removing that code or you could set `UserSettings.lastBoardId = null` here before setting the history.

However, I don't think the code in `boardPage.tsx` is necessary any more. I believe it was there because we didn't have another way to handle it...now we do.

hahmadia replied:
Nice! Good idea. Thanks. Done

sbishel commented on webapp/src/components/emptyCenterPanel.tsx:
```diff
@@ -1,34 +1,170 @@
+    boardTemplate?: Board
+}
+
+const EmptyCenterPanelButton = React.memo((props: ButtonProps) => {
```
Can we just call this a `PanelButton`? There isn't anything requiring it to only be used in the `EmptyCenterPanel` is there?

hahmadia replied:
Sure! Thanks. Done

sbishel reviewed (changes requested):
Hey Hossein. I would rather see the query string `showEmptyCenterPanel` not being used. As mentioned in the comment, I think there are a couple different ways we can do it and be able to remove the property that forces that action. Let me know if you have questions. Thanks.

harshilsharma63 commented on webapp/src/components/emptyCenterPanel.tsx:
```diff
@@ -1,34 +1,170 @@
+        </div>
+    )
+
+    if (Utils.isFocalboardPlugin()) {
```
It's not a nice practice to have the majority of the logic in conditional blocks.

This can be simplified as follows-

```javascript
    if (!Utils.isFocalboardPlugin()) {
        return (
            <div className='EmptyCenterPanel'>
                <div className='Hint'>
                    <FormattedMessage
                        id='EmptyCenterPanel.no-content'
                        defaultMessage='Add or select a board from the sidebar to get started.'
                    />
                </div>
            </div>
        )
    }

    return (...contents of this if block)
```

hahmadia replied:
Smart 💯 Thanks. Done

harshilsharma63 commented on webapp/src/components/emptyCenterPanel.tsx:
```diff
@@ -1,34 +1,170 @@
+                        id='EmptyCenterPanel.plugin.no-content-description'
+                        defaultMessage='Add a board to the sidebar using any of the templates defined below or start from scratch.{lineBreak} Members of "{workspaceName}" will have access to boards created here.'
+                        values={{
+                            workspaceName: <b>{workspace?.title}</b>,
```
A template with no name shows up like this-

<img width="130" alt="Screenshot 2021-10-04 at 1 06 51 PM" src="https://user-images.githubusercontent.com/18575143/135811555-4f346909-3b71-4851-a23e-5d98eb2e431a.png">

Can we add a default text of `Untitled Template` or something similar, same as what we do for boards-

<img width="365" alt="Screenshot 2021-10-04 at 1 07 39 PM" src="https://user-images.githubusercontent.com/18575143/135811681-ac6ac210-147c-4958-8f6c-bf60282e48d8.png">

hahmadia replied:
Done! Thanks

harshilsharma63 commented on webapp/src/components/sidebar/boardTemplateMenuItem.tsx:
```diff
@@ -39,8 +32,49 @@ const addBoardFromTemplate = async (intl: IntlShape, showBoard: (id: string) =>
+                    id='delete'
+                    name={intl.formatMessage({id: 'Sidebar.delete-template', defaultMessage: 'Delete'})}
+                    onClick={async () => {
+                        await mutator.deleteBlock(boardTemplate, 'delete board template')
```
Should there be a confirmation dialogue for this? I know there's another issue where someone implemented it. @asaadmahmood can confirm if we're adopting this design.

hahmadia replied:
If we want to do this, lets create a new ticket for it since it will further delay this ticket. 
I'll skip the implementation for the sake of getting this ticket out to the public sooner.

harshilsharma63 commented on webapp/src/components/emptyCenterPanel.tsx:
```diff
@@ -1,34 +1,170 @@
+            onClick={onClick}
+            className='button'
+        >
+            <i>{buttonIcon}</i>
```
This is sheering the emojis-
<img width="838" alt="Screenshot 2021-10-04 at 1 25 43 PM" src="https://user-images.githubusercontent.com/18575143/135814276-0f9972c2-ca7f-4a30-8b38-61d7d095580f.png">

We should use a span here.

hahmadia replied:
Oooooh Nice! Much better 😄

harshilsharma63 commented:
Awesome feature @hahmadia . WOrks great overall and really love the UI. Some minor things I noticed-

1. The `New Template` button is a bit different from the Figma files. Is that intentional?
<img width="1310" alt="Screenshot 2021-10-04 at 10 58 38 AM" src="https://user-images.githubusercontent.com/18575143/135798741-cf069142-e3b0-4c37-97d9-662b7950633a.png">

2. The workspace switcher select is not width-limited and is breaking for long channel names.
<img width="567" alt="Screenshot 2021-10-04 at 11 00 48 AM" src="https://user-images.githubusercontent.com/18575143/135798865-272e1da0-7298-4be0-93b4-530f9466c3ba.png">
 
3. Long template names break the template button.
<img width="1856" alt="Screenshot 2021-10-04 at 11 02 48 AM" src="https://user-images.githubusercontent.com/18575143/135799025-f9735d3c-24a8-4bda-940c-ca0f08fa199d.png">

4. Some elements don't seem to be themed.
<img width="1317" alt="Screenshot 2021-10-04 at 11 07 32 AM" src="https://user-images.githubusercontent.com/18575143/135799717-48defdd5-1cc7-41a0-8495-1fb7407d3e93.png">

hahmadia commented on webapp/src/components/emptyCenterPanel.scss:
```diff
@@ -11,4 +11,65 @@
+            color: var(--sys-center-channel-color);
+        }
+
+        .bold-grey {
```
Oops! Thanks. Done

hahmadia commented on webapp/src/components/emptyCenterPanel.scss:
```diff
@@ -11,4 +11,65 @@
+            justify-items: center;
+            align-items: center;
+            padding: 16px 32px;
+            border: 1px solid rgba(61, 60, 64, 0.16);
```
Done! Thanks

hahmadia commented:
@harshilsharma63 

Merci for the feedback! 💯 

1. Fixed. Sorry I forgot to do that.
2. Fixed
3. Fixed
4. Border looks better. I'll leave the other up to Asaad.
